### PR TITLE
Pyroclastic anomaly tweak (anti-box)

### PR DESCRIPTION
**Pull request by 0x6273**
Pyroclastic anomaly will now get more unstable at very high temperatures, so if you box it in you will need a freezer to avoid it going critical.

The reason for this change is that before, you could just box it in with glass and not have to deal with it. This is quite boring and Emo has mentioned wanting to change this. See PR #17293.

It's implemented with a `TempReactingAnomalyComponent` that is reusable for other anomalies.

:cl:
- tweak: The pyroclastic anomaly will become more unstable at extreme temperatures, such as when confined to a small space.

### CODE REVIEW

**Comment by deltanedas** [comment]
now you just box it and space the floor

**Comment by EmoGarbage404** [comment]
> now you just box it and space the floor

it has a min and max temp, so you need to manage it both ways.

**Comment by deltanedas** [comment]
ok then scrub plasma into space easy

File: Content.Server/Anomaly/Components/TempReactingAnomalyComponent.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+/// This could later be extended to react only to specific gasses.
+/// </remarks>
+[RegisterComponent]
+public sealed class TempReactingAnomalyComponent : Component
+{
+    /// <summary>
+    /// A list of reactions that happen at certain temperatures.
+    /// </summary>
+    [DataField("reactions")]
+    public List<TempReaction> Reactions = new();
```
Comment (metalgearsloth) [inline Content.Server/Anomaly/Components/TempReactingAnomalyComponent.cs:18]: Should this just be an includedatadef instead.

File: Content.Server/Anomaly/Effects/TempReactingAnomalySystem.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        var indices = _transformSystem.GetGridOrMapTilePosition(uid, tf);
+        var mixture = _atmosphereSystem.GetTileMixture(grid, map, indices);
+
+        if (mixture is null)
+            return;
+
+        foreach (var reaction in tempReactingAnomaly.Reactions)
+        {
+            if (reaction.TempInRange(mixture.Temperature) && TryComp<AnomalyComponent>(uid, out var anomaly))
+            {
```
Comment (metalgearsloth) [inline Content.Server/Anomaly/Effects/TempReactingAnomalySystem.cs:38]: Do the TryComp out of the loop as it's not needed every iteration.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Anomaly/Components/TempReactingAnomalyComponent.cs (hunk 1 above)
Comment (0x6273) [reply Content.Server/Anomaly/Components/TempReactingAnomalyComponent.cs]: Assuming you mean `IncludeDataField` this does not seem to work with `List`. I just changed it to be `required: true`.

**Comment by 0x6273** [comment]
Addressed review, test fail is unrelated to my changes since it happens on master: https://github.com/space-wizards/space-station-14/actions/runs/5287876287/jobs/9568895250

**Comment by EmoGarbage404** [comment]
By now, the pyroclastic anomaly is in a fine place balance-wise.

This change is unintuitive and not really in line with the other anomalies. I don't feel like this is the way to go about doing this.

**Comment by AJCM-git** [comment]
Another better would probably be making high pressure/temperature always damage windows.

**Comment by 0x6273** [comment]
> Another better would probably be making high pressure/temperature always damage windows.

I like this idea, but I think people will just use walls if this is only for windows. I guess those could also be damaged if temps get high enough to melt steel. I'm not sure how to code this in a good way since I imagine having every wall/window in the game check surrounding temps would probably tank performance.

**Comment by AJCM-git** [comment]
> I like this idea, but I think people will just use walls if this is only for windows. I guess those could also be damaged if temps get high enough to melt steel. I'm not sure how to code this in a good way since I imagine having every wall/window in the game check surrounding temps would probably tank performance.

Well if they box with walls then they cant blast it with particles to avoid a meltdown. Tho pressure should definitively damage walls, idk enough about atmos code to tell you a way to do this efficiently, the only lead i would have for something like that is taking a look at how atmos superconductivity works (its an atmos CVAR that makes heat transportable through walls which **may** or may not help you arrive at a solution)

**Pull request closed without merging**